docs/docs/reference/experimental/cc.md

```diff
@@ -0,0 +1,715 @@
+  def recur(i: Int): {gen} LzyList[A] =
+    if i == n then LzyNil
+    else gen(i) #: recur(i + 1)
+  recur(0)
+```
+Here is a use of `tabulate`:
+```scala
+class LimitExceeded extends Exception
+def squares(n: Int)(using ct: CanThrow[LimitExceeded]) =
+  tabulate(10) { i =>
+    if i > 9 then throw Ex1()
+    i * i
+  }
+```
+The inferred result type of `squares` is `{ct} LzyList[Int]`, i.e it is a lazy list of
+`Int`s that can throw the `LimitExceeded` exception when it is elaborated by calling `tail`
+one or more times.
+
+Here are some further extension methods for mapping, filtering, and concatenating lazy lists:
+```scala
+extension [A](xs: {*} LzyList[A])
+  def map[B](f: A => B): {xs, f} LzyList[B] =
+    if xs.isEmpty then LzyNil
+    else f(xs.head) #: xs.tail.map(f)
+
+  def filter(p: A => Boolean): {xs, p} LzyList[A] =
+    if xs.isEmpty then LzyNil
+    else if p(xs.head) then xs.head #: xs.tail.filter(p)
+    else xs.tail.filter(p)
+
+  def concat(ys: {*} LzyList[A]): {xs, ys} LzyList[A] =
+    if xs.isEmpty then ys
+    else xs.head #: xs.tail.concat(ys)
+
+  def drop(n: Int): {xs} LzyList[A] =
+    if n == 0 then xs else xs.tail.drop(n - 1)
+```
+Their capture annotations are all as one would expect:
+
+ - Mapping a lazy list produces a lazy list that captures the original list as well
+   as the (possibly impure) mapping function. Of course, it would also be possible to
+   pass a concrete function argument that is pure.
+ - Filtering a lazy list produces a lazy list that captures the original list as well
+   as the (possibly impure) filtering predicate.
+ - Concatenating two lazy lists produces a lazy list that captures both arguments.
+ - Dropping elements from a lazy list gives a safe approximation where the original list is captured in the result. In fact, it's only some suffix of the list that is retained at run time, but our modelling identifies lazy lists and their suffixes, so this additional knowledge would not be useful.
+
+Of course the function passed to `map` or `filter` could also be pure. After all, `A -> B` is a subtype of `{*} A -> B` which is the same as `A => B`. In that case, the pure function
+argument will _not_ show up in the result type of `map` or `filter`. For instance:
+```scala
+val xs = squares(10)
+val ys: {xs} LzyList[Int] = xs.map(_ + 1)
+```
+The type of the mapped list `ys` has only `xs` in its capture set. The actual function
+argument does not show up since it is pure. Likewise, if the lazy list
+`xs` was pure, it would not show up in any of the method results.
+This demonstrates that capability-based
+effect systems with capture checking are naturally _effect polymorphic_.
+
+This concludes our example. It's worth mentioning that an equivalent program defining and using standard, strict lists would require no capture annotations whatsoever. It would compile exactly as written now in standard Scala 3. This fact is probably one of the greatest plus points of our approach to capture checking.
```

Comment:
Hello Martin, I am probably not suited to do a review. But as a potential user of this effect system I would like to ask a clarifying question. I don't fully understand this last paragraph. Is it because the standard list is strict, or because it is defined without any capture checking that no annotations are necessary?

If I would like to use a strict list but also track whether something has effects, is that possible? And if so, would that require a new strict data structure or in another way?

Reply:
It is because it is strict. You get the capture checking for free. Essentially, `=>` already means "can capture anything" and since in a strict list side effecting operations are not retained in the result, there's nothing captured at all. A strict list could of course capture side effecting closures in its elements but then tunnelling kicks and we don't need to annotate anything there either.

Reply:
Thanks for the question, it will hopefully help me to make that point clearer!

Reply:
No problem, and thanks for the quick response. I think I understand it a little better now. Effects are only captured if they are not executed yet, so either in closures or lazy data structures.

Reply:
Another point to make is that in order to call `squares` you must have the canthrow capability in scope, so you would already expect an exception to be thrown possibly when calling it. The reason that if `squares` returned a strict list it would not need to track canthrow in its return type is because at that point the new list is already created and no exceptions were thrown, so that list can no longer throw exceptions (but the lazylist can still throw if it escapes and you evaluate the tail elements).

Reply:
The current stdlb will have to be modified, to account for delayed effects in views or iterators. Hopefully, the scope of the modifications will be not too large. If things work out as hoped for, we might be able to roll that into the stdlib the next time we update it. But probably this would be a 3.x version of the library, it might be difficult to maintain compatibility with 2.13.